tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -1092,6 +1094,93 @@ LogicalResult ConvertTFLConv2DOp::matchAndRewrite(
+      tfl_conv3d_op.input().getType().dyn_cast<RankedTensorType>();
+  RankedTensorType filter_type =
+      tfl_conv3d_op.filter().getType().dyn_cast<RankedTensorType>();
+  ShapedType output_type =
+      tfl_conv3d_op.getResult().getType().dyn_cast<ShapedType>();
+
+  if (!input_type | !filter_type || !output_type) {
+    return rewriter.notifyMatchFailure(
+        op, "input/filter/output are not all a ranked tensor");
+  }
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool filter_is_qtype =
+      filter_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::QuantizedType>();
+
+  if ((input_is_qtype != filter_is_qtype) ||
+      (input_is_qtype != output_is_qtype)) {
+    return rewriter.notifyMatchFailure(op,
+                                       "input/filter/output tensor should be "
+                                       "all quantized or all floating-point");
+  }
+
+  int64_t stride_d = tfl_conv3d_op.stride_d();
+  int64_t stride_h = tfl_conv3d_op.stride_h();
+  int64_t stride_w = tfl_conv3d_op.stride_w();
+  ArrayAttr stride = rewriter.getI64ArrayAttr({stride_d, stride_h, stride_w});
+
+  int64_t dilation_d = tfl_conv3d_op.dilation_d_factor();
+  int64_t dilation_h = tfl_conv3d_op.dilation_h_factor();
+  int64_t dilation_w = tfl_conv3d_op.dilation_w_factor();
+  ArrayAttr dilation =
+      rewriter.getI64ArrayAttr({dilation_d, dilation_h, dilation_w});
+
+  Value unquantized_bias = tfl_conv3d_op.bias();
+  if (!unquantized_bias.getType().dyn_cast<RankedTensorType>()) {
+    // The bias may actually be typed "None" which has no value. TOSA requires
+    // bias to be an array of output_channel_count values, so create a constant
+    // of the appropriate number and type of zeros.
+    auto bias_dim = filter_type.getShape().back();
+    RankedTensorType bias_type =
+        RankedTensorType::get({bias_dim}, filter_type.getElementType());
+    auto bias_attr = rewriter.getZeroAttr(bias_type);
+    unquantized_bias = CreateOpAndInfer<tosa::ConstOp>(
+        rewriter, op->getLoc(), bias_type, bias_attr.cast<ElementsAttr>());
+  }
+
+  Type bias_ety =
+      unquantized_bias.getType().cast<ShapedType>().getElementType();
+  llvm::Optional<Value> a1_conv3d_op = convertConv3DCommon(
+      rewriter, op, output_type.clone(bias_ety), tfl_conv3d_op.input(),
+      tfl_conv3d_op.filter(), unquantized_bias, stride, dilation,
+      tfl_conv3d_op.padding().str(), StringRef("NDHWC"));
+
+  if (!a1_conv3d_op) return failure();
+
+  Value conv3d_output;
+  if (input_is_qtype) {
+    conv3d_output =
+        buildRescaleOpConvOutput(rewriter, op, a1_conv3d_op.value(),
+                                 input_type, filter_type, output_type);
+  } else {
+    conv3d_output = a1_conv3d_op.value();
+  }
+
+  auto fused_activation_fn = tfl_conv3d_op.fused_activation_functionAttr();
+  if (fused_activation_fn) {
```

Comment:
To reduce scope/lifeness, 

if (auto fused_activation_fn = ...) { ... }

Reply:
Done.